Label/Billboard/Point clustering

bagnell opened the pull request:
Adds label, billboard, and point clustering. There is a new object on each data source called `clustering`. This has properties `enabled`, `pixelRange` and `minimumClusterSize`. There is also `clusterEvent` that is raised for every new cluster. The parameters to the callback are the array of entities being clustered and an entity that will be drawn for the cluster. The `label` and `billboard` properties for the second entity parameter are the only things considered when drawing the cluster.

TODO:
- [x] Fix existing tests
- [x] Add new tests
- [x] Update doc
- [x] Update `CHANGES.md`
- [x] Update `LICENSE.md` for `kdbush`
- [ ] Submit issues for deprecated function parameters

bagnell commented:
If you run the Sandcastle and switch to 2D/CV, all of the labels and billboards are in the wrong position. That isn't a problem with the new code. The same problem exists in master with the KML Sandcastle example. I think the problem is in the clamp-to-ground code.

pjcozzi commented:
Note that this is being merged into `cluster-master` as we don't anticipate this making 1.25 despite thinking otherwise offline.

pjcozzi commented:
Can the Sandcastle example have clustering on by default?

pjcozzi commented:
Is it hard for the default styling to fade in/out clusters since they has become standard practice?

pjcozzi commented:
Do the defaults cluster aggressively enough?

![image](https://cloud.githubusercontent.com/assets/782098/18029288/e4e3b134-6c61-11e6-9648-7aaa4f110d68.png)

pjcozzi commented:
In the Sandcastle example, can "Toggle Custom Styling" force a redraw?  Otherwise, it is not obvious that anything changed.

pjcozzi commented:
Can the default styling add commas?

pjcozzi commented on Source/DataSources/EntityCluster.js:
```diff
@@ -0,0 +1,775 @@
+        width += pixelRange * 0.5;
+        height += pixelRange * 0.5;
+
+        return new BoundingRectangle(x, y, width, height);
```
Should this use a `result` parameter?

pjcozzi commented on Source/DataSources/EntityCluster.js:
```diff
@@ -0,0 +1,775 @@
+        return point.coord.y;
+    }
+
+    function getLabelBoundingBox(label, coord, pixelRange) {
```
Can this go on `Label` as a `@private` function?

pjcozzi commented on Source/DataSources/EntityCluster.js:
```diff
@@ -0,0 +1,775 @@
+        return new BoundingRectangle(x, y, width, height);
+    }
+
+    function getBillboardBoundingBox(billboard, coord, pixelRange) {
```
Same two comments for this function and `getPointBoundingBox`.

pjcozzi commented on Source/DataSources/EntityCluster.js:
```diff
@@ -0,0 +1,775 @@
+        }
+    }
+
+    function createDeclutterCallback(entityCluster) {
```
Is it possible to move this clustering algorithm into `Core` as a `@private` class/function?

pjcozzi replied:
Or `Scene` I suppose.

bagnell replied:
Yes, but I prefer not to. It's highly coupled to this class.

pjcozzi commented on Source/DataSources/EntityCluster.js:
```diff
@@ -0,0 +1,775 @@
+        };
+    }
+
+    function EntityCluster(options) {
```
This should go at the top of the file.  Relying on function hoisting is OK.

pjcozzi commented on Source/DataSources/LabelVisualizer.js:
```diff
@@ -67,24 +69,29 @@ define([
         }
         //>>includeEnd('debug');
 
+        if (!entityCluster.minimumClusterSize) {
```
Submit an issue for this or add it to this PR's tasklist.

pjcozzi replied:
Same comment throughout, of course.

pjcozzi commented on Source/Scene/Label.js:
```diff
@@ -701,6 +703,27 @@ define([
             }
+        },
+
+        _clusterRender : {
```
This doesn't need an underscore; it can just be `@private`.  Not sure that I would use "render" here though since we never use in this context.  Perhaps just `clusterShow`.

pjcozzi commented on Source/ThirdParty/kdbush.js:
```diff
@@ -0,0 +1,216 @@
+(function(f){if(typeof exports==="object"&&typeof module!=="undefined"){module.exports=f()}else if(typeof define==="function"&&define.amd){define([],f)}else{var g;if(typeof window!=="undefined"){g=window}else if(typeof global!=="undefined"){g=global}else if(typeof self!=="undefined"){g=self}else{g=this}g.kdbush = f()}})(function(){var define,module,exports;return (function e(t,n,r){function s(o,u ...
```
Update LICENSE.md or add it to the tasklist.

bagnell commented:
>Do the defaults cluster aggressively enough?

The defaults do. I had to modify them in this example because the KML isn't very good. There are a bunch of billboards and labels at the same position with slightly different text. I saw a maximum of three overlapping so I had to change the options for this demo to not cluster as aggressively.

bagnell commented:
@pjcozzi This is ready for another look.
@mramato Can you review this as well when you get a chance?

pjcozzi commented:
Can the Sandcastle example cluster onload (in general, can data sources do this?).  It should look like this:

![image](https://cloud.githubusercontent.com/assets/782098/18225846/de5b21a2-71cb-11e6-88bf-94087b97efc9.png)

Not this:

![image](https://cloud.githubusercontent.com/assets/782098/18225852/f280c452-71cb-11e6-9a6d-9b3af8ba37c2.png)

pjcozzi commented:
In the Sandcastle example, can we default to the custom styling so it gives a better first impression?

pjcozzi commented:
Could be worth spending a few minutes to see how the custom styling in the Sandcastle example looks when a circle primitive shows the cluster size (does the callback include enough info for this?)

pjcozzi commented:
> Could be worth spending a few minutes to see how the custom styling in the Sandcastle example looks when a circle primitive shows the cluster size (does the callback include enough info for this?)

You might only want to do this for some zoom levels, i.e., not those when zoomed too far out.

pjcozzi commented:
A visible label can become clustering when zoomed in and then visible again when moved in farther.  Is there anything we can do about that?

See "Mountain Horse Station" below:

![ok](https://cloud.githubusercontent.com/assets/782098/18225980/47d624ae-71ce-11e6-9637-34e7f9437c34.gif)

pjcozzi commented on Apps/Sandcastle/gallery/Clustering.html:
```diff
@@ -0,0 +1,158 @@
+
+    function subscribeParameter(name) {
+        Cesium.knockout.getObservable(viewModel, name).subscribe(
+                function(newValue) {
```
Extra indent?

pjcozzi commented on Apps/Sandcastle/gallery/Clustering.html:
```diff
@@ -0,0 +1,158 @@
+    var removeListener;
+
+    var pinBuilder = new Cesium.PinBuilder();
+    var pin20 = pinBuilder.fromText('20+', Cesium.Color.RED, 48).toDataURL();
```
Given how many labels there are, I think this would show better with a few levels above "20+"

pjcozzi commented on Source/Scene/Billboard.js:
```diff
@@ -1149,6 +1172,49 @@ define([
+     *
+     * @private
+     */
+    Billboard.getScreenSpaceBoundingBox = function(billboard, screenSpacePosition, result) {
```
Can you add unit tests for all the new `getScreenSpaceBoundingBox`?

Also, although I am all for keeping the API surface area small, have folks asked for these functions before?  It could be worth adding public `prototype` functions for the public API?

bagnell replied:
I don't think so. @hpinkos, have you seen anyone ask for something like this on the forum?

pjcozzi commented:
That's all my comments.  This is a hugely welcomed feature!

> @mramato Can you review this as well when you get a chance?

Could you please review (and ideally merge) this week?

bagnell commented:
All of @pjcozzi 's comment have been addressed. Just waiting on feedback from @mramato.

mramato commented:
I didn't see this mentioned, but morphing performance is severely degraded in the example.  It takes several seconds for the morph to even start with the default view/example.

pjcozzi commented:
+1 on morphing issue; I thought I was just having browser issues, but I saw that too.

mramato commented:
This seems completely broken for time-dynamic data, did you test it at all?  I have a file you can use I'll send you offline (too big to post here).

mramato commented on Apps/Sandcastle/gallery/Clustering.html:
```diff
@@ -0,0 +1,172 @@
+        } else {
+            removeListener = dataSource.clustering.clusterEvent.addEventListener(function(cluster, entity) {
+                entity.label = undefined;
+                entity.billboard = {};
```
You probably want to set the origin to bottom here so the pins point is at the location of the cluster instead of at the center of the billboard.

colek42 commented:
I am getting an error ```...Uncaught TypeError: Cannot read property 'get' of undefined```

this is the offending function
```
 function getBoundingBox(item, coord, pixelRange, entityCluster, result) {
        if (defined(item._labelCollection)) {
            result = Label.getScreenSpaceBoundingBox(item, coord, result);
        } else if (defined(item._billboardCollection)) {
            result = Billboard.getScreenSpaceBoundingBox(item, coord, result);
        } else if (defined(item._pointPrimitiveCollection)) {
            result = PointPrimitive.getScreenSpaceBoundingBox(item, coord, result);
        }

        expandBoundingBox(result, pixelRange);

        if (!defined(item._labelCollection) && defined(item.id._label)) {
            var labelIndex = item.id._labelIndex;
            var label = entityCluster._labelCollection.get(labelIndex); <---
            var labelBBox = Label.getScreenSpaceBoundingBox(label, coord, labelBoundingBoxScratch);
            expandBoundingBox(labelBBox, pixelRange);
            result = BoundingRectangle.union(result, labelBBox, result);
        }

        return result;
    }
```

colek42 commented:
When an entity is added to a datasource the clustering is not updated until the camera position is moved.

pjcozzi commented:
> When an entity is added to a datasource the clustering is not updated until the camera position is moved. 

This is also noticeable in the Sandcastle example, see https://github.com/AnalyticalGraphicsInc/cesium/pull/4240#issuecomment-244553755 above.

bagnell commented:
>I didn't see this mentioned, but morphing performance is severely degraded in the example. It takes several seconds for the morph to even start with the default view/example.

The performance degradation isn't caused by this branch. Try the KML example in master.

bagnell commented:
>This seems completely broken for time-dynamic data, did you test it at all? I have a file you can use I'll send you offline (too big to post here).

I need to ad this to the documentation. Clustering is only supported for static data. Otherwise, this is ready for another review.

pjcozzi commented:
@bagnell what is left here?  Just

>  [ ] Fix labels clamped to ground or screen space position in 2D/CV?

Does this need another review?

bagnell commented:
@pjcozzi No, that is a problem with the with clamping to terrain so I removed it from the list. There was one minor code change and updated Sandcastle example and doc. Waiting for @mramato's code review.

mramato commented:
Yeah, clamping on terrain is generally busted in master.  We should really fix that soon (in a separate PR).

pjcozzi commented:
> Waiting for @mramato's code review.

@mramato do you please on reviewing this soon so it can make the 1.26 release?

> Yeah, clamping on terrain is generally busted in master. We should really fix that soon (in a separate PR).

Please submit an issue if there isn't already one.  Could be good for the bug bash or sooner.

mramato commented:
I will try to get to this soon.

#4062 and #4300 should cover the clamping issues

mramato commented:
1. I'm still seeing problems where billboards cluster and then uncluster as I zoom farther out.
2. Is clustering deterministic?  If I maintain a constant direction and simply move the camera closer/further away, I see different results when revisiting the same distance.  This is easy to reproduce by loading the cluster demo, zooming to an individual entity, and then zooming the camera in and out repeatedly.
3. Similarly, enable clustering, get a view with lots of clusters, toggle it off, then toggle back on.  The clusters are now different.  Why?

mramato commented:
FYI, while I was testing the above, I was specifically locked onto `Fort Monmouth` in Jersey.

mramato commented:
I get that this isn't supposed to work with time-dynamic data, but the current behavior is to try and cluster and then crash.  Why aren't we just ignoring time-dynamic positions?

mramato commented:
By the way, it actually seems to work really well on time-dynamic data until it crashes.  So maybe the crash is simply pointing out a lower level issue.

```
An error occurred while rendering.  Rendering has stopped.
undefined
TypeError: Cannot read property '_label' of undefined
TypeError: Cannot read property '_label' of undefined
    at getBoundingBox (http://localhost:8080/Source/DataSources/EntityCluster.js:130:63)
    at EntityCluster._cluster (http://localhost:8080/Source/DataSources/EntityCluster.js:383:24)
    at EntityCluster.update (http://localhost:8080/Source/DataSources/EntityCluster.js:758:18)
    at PrimitiveCollection.update (http://localhost:8080/Source/Scene/PrimitiveCollection.js:366:27)
    at updatePrimitives (http://localhost:8080/Source/Scene/Scene.js:2185:27)
    at executeCommandsInViewport (http://localhost:8080/Source/Scene/Scene.js:2101:9)
```

mramato commented:
ICRF appears to work, but has the same (appearance of) non-deterministicness.

colek42 commented:
We don't mind the clusters being out of sync.  Most of the movement in our time dynamic data is sub pixel at the zoom levels we will be using it at.  We would appreciate a recluster() function that we can call to ensure that the clusters are updated once in a while.  Clustering is one of the biggest feature requests that we get.  Thanks for working on it.

mramato commented:
Have we thought about allowing cluster selection to see what's in the cluster?  Not saying it needs to be part of this PR, but no default handling for a cluster seems incomplete.  We've talked about multiselection in the past and having multiple tabs in the InfoBox to do it. (or we could show the clusters description, which if set, currently does nothing).

mramato commented on Apps/Sandcastle/gallery/Clustering.html:
```diff
@@ -0,0 +1,174 @@
+        </tr>
+        </tbody>
+    </table>
+    <div id="toggleClustering"></div>
```
Can we make these checkboxes, the toggle button is confusing because you don't know what state it's in.

bagnell commented:
@mramato We discussed the issues with how clustering works offline. I submitted changes to fix time-dynamic data like we discussed as well and fixed the issue when removing points. This is ready for another review.

mramato commented:
Using the cluster demo:
1. Switch to 2D (you'll get some weird billboard flashing)
2. Disable clustering
3. Everything is broken

![image](https://cloud.githubusercontent.com/assets/1495196/18913610/a6c8a072-8556-11e6-8e01-95814afd215a.png)

Similar problems in Columbus View

mramato commented:
To experiment with custom styling, I tried making a point where the size depends on the size of the cluster (instead of a billboard).  Is there a reason this doesn't work?

```javascript
removeListener = dataSource.clustering.clusterEvent.addEventListener(function(cluster, entity) {
    entity.label = undefined;
    entity.point = {
        pixelSize : cluster.length
    };
});
```

bagnell commented:
>Using the cluster demo:
1. Switch to 2D (you'll get some weird billboard flashing)
2. Disable clustering
3. Everything is broken

This was mentioned above. The labels are clamped to the ground and its broken. Open the KML example that uses the same KML file and you'll see the same behavior.

bagnell commented:
>To experiment with custom styling, I tried making a point where the size depends on the size of the cluster (instead of a billboard). Is there a reason this doesn't work?

The code that adds the cluster is here: https://github.com/AnalyticalGraphicsInc/cesium/pull/4240/files#diff-3ca360d29d7d55bfdc7322d4e18721e7R157 Its pretty limited because I wanted to keep it close to the cluster position and limit the size so it doesn't overlap other clusters. I'm open to any suggestions here.

mramato commented:
> This was mentioned above. The labels are clamped to the ground and its broken. Open the KML example that uses the same KML file and you'll see the same behavior.

That breakage is worst than I thought then.  We should fix this ASAP.  I'll mark those issues as next release.

bagnell commented:
@mramato I updated the custom styling like we discussed offline. This is ready for another look.

mramato commented:
The doc needs to be updated, but then I think this is good to go.

bagnell commented:
@mramato The doc is updated. This is ready.

mramato commented:
Thanks @bagnell!
